HIP/hip-xxx.md

```diff
@@ -0,0 +1,921 @@
+                    "0000000000000000000000000000000000000000000000000000000000000080"
+                ],
+                "storage": {
+                    "0x0000000000000000000000000000000000000000000000000000000000000000": "0x0000000000000000000000000000000000000000000000000000000000000014"
+                },
+                "reason": null
+            },
+            {
+                "pc": 544,
+                "op": "PUSH1",
+                "gas": 4414,
+                "gasCost": 3,
+                "depth": 1,
+                "stack": null,
+                "memory": [
+                    "0x0000000000000000000000000000000000000000000000000000000000000000",
+                    "0000000000000000000000000000000000000000000000000000000000000000",
+                    "0000000000000000000000000000000000000000000000000000000000000080"
+                ],
+                "storage": {
+                    "0x0000000000000000000000000000000000000000000000000000000000000000": "0x0000000000000000000000000000000000000000000000000000000000000014"
+                },
+                "reason": null
+            },
+            {
+                "pc": 546,
+                "op": "SLOAD",
+                "gas": 4411,
+                "gasCost": 2100,
+                "depth": 1,
+                "stack": null,
+                "memory": [
+                    "0x0000000000000000000000000000000000000000000000000000000000000000",
+                    "0000000000000000000000000000000000000000000000000000000000000000",
+                    "0000000000000000000000000000000000000000000000000000000000000080"
+                ],
+                "storage": {
+                    "0x0000000000000000000000000000000000000000000000000000000000000000": "0x0000000000000000000000000000000000000000000000000000000000000014"
+                },
+                "reason": null
+            },
+            [...]
+        ]
+}
+```
+
+For providing output formatted by `opcodeLogger`, the Mirror node will re-execute the transaction using the state from the `contract_state_changes` sidecars produced by the consensus nodes. In this way, we can have a track on all of the storage/memory information and the entire trace of opcodes that were executed during the replay.
+
+⚠️ The output generated from the opcodeLogger is very verbose. The /traceOpcodes endpoint will be disabled by default and can be enabled by users of the Hedera Local node or users running their own Mirror Nodes.
+
+### **Performance considerations**
+
+Since the output generated by this tracer might be quite large, we could add a `flag` for enabling different type of tracers, based on the usage and setup of the Mirror Archive Node. This would give us control and flexibility. A user would be able to enable `opcodeLogger` in a local environment and get the full response data. The default tracer enabled by default on production, could be the `callTracer`.
```

Comment:
It would be nice to have worst case numbers to define "large". e.g how many max opcodes can be returned and the largest size of any particular entry with all flags enabled. I believe Danno mentioned in a call but I can't recall and would be good to have here so we can make informative decisions.

Reply:
I ran a transaction with 1024 nested calls (the maximum allowed by the EVM) with all flags enabled. The output size for the opcodes logger info was 432MB, which seems quite a lot. For a single nested call the size was around 500 KB and the max pc counter is 1895. So for a main + nested transaction we get around 1895 op codes.

The memory of a given contract is limited to 2^256 slots of 32 bytes of entries.
The storage I believe has no limit and you store theoretically even store GB of data if you are ready to pay for this. So it seems we have no control or predictability of the storage size.

I suggest allowing the `/opcodes` endpoint to work only for local environments and additionally add pagination logic (again only locally). Pagination would mean executing the transaction again for each pagination request and fetching different portion of the output, so it won't be so efficient in terms of computation power, but could save us output size.

@Nana-EC , @steven-sheehy what are your thoughts?

Reply:
What would be the pagination key if we added it? pc?

I assume there's a way to halt execution after reaching the page size to optimize earlier pages, but not continuing from there on the next page?

Since it's all local we could just not add pagination now. Then later if it's necessary we could add it in a backwards compatible way via an extra query param and a `links` section. Also, if we decide to enable this in an environment like testnet we could also consider limiting requests based upon the gas in the db. We query the gas and if it's within some reasonable range query for the opcodes and return it. Otherwise we return an error or partial data

Reply:
I agree with you. For now we can postpone the idea of pagination and come back in case of need. 

We can additionally limit requests based on the potential `gasCharged` value we might introduce inside the importer. This would be a good measurement of how much op codes would be executed and relatively how big the response size would be.